### DISCUSSION TITLE
Node-level Classification using ECCConv

### DISCUSSION THREAD

**Discussion by melheithem** [danielegrattarola/spektral#217] [Q&A] [answered]
Hi,

I'm trying to do node classification using Spektral on a dataset that has edge features (I have different types of edges). I found an example for graph-level classification using ECCConv (examples / graph_prediction / qm9_ecc.py), is there a similar example for classification at node level?

Thanks,

**Comment by danielegrattarola** [comment]
Hi, 

no, there are no examples for that use case currently, but you can easily adapt the graph-level classification examples to node prediction by removing the global pooling layer from the model. That will leave you with a model based on ECC and that will output predictions for all nodes. 

Cheers

**Reply by melheithem** [reply]
Hi,

Thanks for your answer, I tried to adapt the graph level classification by removing the global pooling layer as below : 

```
X_in = Input(shape=(None,NF))
A_in = Input(shape=(None,None)) #, sparse=True)
E_in = Input(shape=(None,None,EF))

X_1 = ECCConv(32, activation="relu")([X_in, A_in, E_in]) 
X_2 = ECCConv(n_out, activation="softmax")([X_1, A_in, E_in]) 

model = Model(inputs=[X_in, A_in, E_in], outputs=X_2)
opt = Adam(lr=learning_rate)
model.compile(optimizer=opt, loss="categorical_crossentropy")
model.fit(
    loader_tr.load(),
    steps_per_epoch=loader_tr.steps_per_epoch,
    validation_data=loader_va.load(),
    validation_steps=loader_va.steps_per_epoch,
    epochs=epochs,
    callbacks=[EarlyStopping(patience=es_patience, restore_best_weights=True)],
)
```
I got the following error :

```
InvalidArgumentError:  Incompatible shapes: [1148] vs. [6400]
	 [[node gradient_tape/model/ecc_conv_1/Maximum (defined at <ipython-input-8-c3f52bd46ea8>:7) ]] [Op:__inference_train_function_1256]

Errors may have originated from an input operation.
Input Source operations connected to node gradient_tape/model/ecc_conv_1/Maximum:
 model/ecc_conv_1/strided_slice_2 (defined at /home/emohammedi/.local/lib/python3.6/site-packages/spektral/layers/convolutional/ecc_conv.py:172)

Function call stack:
train_function
```
Regards,

**Reply by danielegrattarola** [reply]
Can you give me more details about the dataset that you are using when you get this crash??
Cheers

**Reply by melheithem** [reply]
Hi,

It's a custom dataset which I created for a network configuration issue. It contains one graph with multiple nodes. Each node contains 12 features, and each edge contains 10 features which refer to common parameters (VRFs - Virtual Routing and Forwarding) between nodes.

For the first tests, I generated 3 datasets for the train (80 nodes), validation (10 nodes) and test (10 nodes). For now, I am generating small datasets just to test my model (my code). Here is my code to generate the datasets :

![image](https://user-images.githubusercontent.com/27120857/117225338-690c7300-ae12-11eb-9355-656b689186ea.png)

Best regards,
Heithem

**Comment by danielegrattarola** [comment, marked as answer]
I see. 

Can you try to update the inputs of your model to:

```py
X_in = Input(shape=(NF,))
A_in = Input(shape=(None,)) #, sparse=True)
E_in = Input(shape=(None,EF))
```

Or you can even remove the inputs altogether and implement your model via subclassing. 

Cheers

**Reply by melheithem** [reply]
Hi,

I have updated the inputs but still get the same error. I will try to implement the model via subclassing and keep you posted.

Thank you !

**Reply by melheithem** [reply]
Hi, 

I tried to implement the model via subclassing as below : 

```
class Net(Model):
    def __init__(self, n_out, **kwargs):
        super().__init__(**kwargs)
        self.conv1 = ECCConv(32, activation="relu")
        self.conv2 = ECCConv(n_out, activation="softmax")

    def call(self, inputs):
        x, a, e = inputs
        x = self.conv1([x, a, e])
        output = self.conv2([x, a, e])

        return output


model = Net(n_out=n_out)
opt = Adam(lr=learning_rate)
loss_fn = CategoricalCrossentropy()
model.compile(optimizer=opt, loss=loss_fn) 

model.fit(
    loader_tr.load(),
    steps_per_epoch=loader_tr.steps_per_epoch,
    validation_data=loader_va.load(),
    validation_steps=loader_va.steps_per_epoch,
    epochs=epochs,
    callbacks=[EarlyStopping(patience=es_patience, restore_best_weights=True)],
)
```
I still get the same error : 

```
InvalidArgumentError:  Incompatible shapes: [1207] vs. [6400]
	 [[node gradient_tape/net/ecc_conv_1/Maximum (defined at <ipython-input-8-c3f52bd46ea8>:7) ]] [Op:__inference_train_function_1077]

Errors may have originated from an input operation.
Input Source operations connected to node gradient_tape/net/ecc_conv_1/Maximum:
 net/ecc_conv_1/strided_slice_2 (defined at /home/emohammedi/.local/lib/python3.6/site-packages/spektral/layers/convolutional/ecc_conv.py:172)

Function call stack:
train_function
```

I also tried to test with the functional model using a custom function for training, I got a warning message and another error : 

```
Fitting model
WARNING:tensorflow:Model was constructed with shape (None, None, 10) for input KerasTensor(type_spec=TensorSpec(shape=(None, None, 10), dtype=tf.float32, name='input_3'), name='input_3', description="created by layer 'input_3'"), but it was called on an input with incompatible shape (None, 10).

...

ValueError: in user code:

    <ipython-input-10-35d39475a874>:8 train_step  *
        loss = loss_fn(target, predictions)
    /home/emohammedi/.local/lib/python3.6/site-packages/tensorflow/python/keras/losses.py:152 __call__  **
        losses = call_fn(y_true, y_pred)
    /home/emohammedi/.local/lib/python3.6/site-packages/tensorflow/python/keras/losses.py:256 call  **
        return ag_fn(y_true, y_pred, **self._fn_kwargs)
    /home/emohammedi/.local/lib/python3.6/site-packages/tensorflow/python/util/dispatch.py:201 wrapper
        return target(*args, **kwargs)
    /home/emohammedi/.local/lib/python3.6/site-packages/tensorflow/python/keras/losses.py:1537 categorical_crossentropy
        return K.categorical_crossentropy(y_true, y_pred, from_logits=from_logits)
    /home/emohammedi/.local/lib/python3.6/site-packages/tensorflow/python/util/dispatch.py:201 wrapper
        return target(*args, **kwargs)
    /home/emohammedi/.local/lib/python3.6/site-packages/tensorflow/python/keras/backend.py:4833 categorical_crossentropy
        target.shape.assert_is_compatible_with(output.shape)
    /home/emohammedi/.local/lib/python3.6/site-packages/tensorflow/python/framework/tensor_shape.py:1134 assert_is_compatible_with
        raise ValueError("Shapes %s and %s are incompatible" % (self, other))

    ValueError: Shapes (11,) and (None, 11) are incompatible
```

The input and output shapes that I am using are : 

![image](https://user-images.githubusercontent.com/27120857/117663105-9190c600-b1a0-11eb-8d66-557e6561b8d3.png)


Thanks,
Heithem

**Reply by danielegrattarola** [reply]
This is weird.
Are you using a DisjointLoader? If so, are you setting node_level=True?

Cheers

**Reply by melheithem** [reply]
No, I am using a SingleLoader. Should I use a DisjointLoader?

Thanks,

**Reply by danielegrattarola** [reply]
No, I was just trying to find possible problems. 

Ok so, here's a script that should do what you are trying to do: 

```python
import numpy as np
import scipy.sparse as sp
from spektral.data import Graph, Dataset, SingleLoader
from spektral.layers import ECCConv
from tensorflow.keras.models import Model

N = 10
F = 3
S = 4
n_out = 5


class MyDataset(Dataset):
    def read(self):
        return [
            Graph(
                x=np.ones((N, F)), 
                a=sp.coo_matrix(np.ones((N, N))),
                e=np.ones((N, N, S)), 
                y=np.ones((N, n_out))
            )
        ]


class Net(Model):
    def __init__(self):
        super().__init__()
        self.conv = ECCConv(n_out, activation="relu")

    def call(self, inputs):
        x, a, e = inputs
        x = self.conv([x, a, e])

        return x


data = MyDataset()
loader = SingleLoader(data)
model = Net()
model.compile('adam', 'mse')
model.fit(loader.load(), steps_per_epoch=loader.steps_per_epoch)

```

This should throw no errors. Can you verify that your data and/or training script match this general format?

**Reply by melheithem** [reply]
I tried using your sample data (I put the same shapes as my dataset) with my model. I haven't changed anything in my training script and it works. It's weird because I am generating my dataset the same way as your script. 

Surely there is an error somewhere on my dataset. I can see more clearly now. I will try to solve it. 

I will keep you posted.

Thank you so much !

**Reply by danielegrattarola** [reply]
Good luck :D
Let me know if you solve it

**Reply by melheithem** [reply]
I think I found the error, for a dataset with 20 edges and 10 nodes, I get the error : `Incompatible shapes: [20] vs. [100]` and for a dataset with 2 edges and 5 nodes, I get :  `Incompatible shapes: [2] vs. [25]`. In your sample data, you have all possible edges by using :  `a = np.ones((N,N))`. I tried to do the same with my Dataset and it works well. I think the incompatible shapes are [n_edges] vs. [N*N].

**Reply by danielegrattarola** [reply]
You are correct, I see the problem now and it was obvious in retrospective. 
 
Short fix: change the edge features matrix so that it has shape `(n_edges, n_edge_features)`. This is achieved by stacking the edge attributes sorted by the lexicographic order of the adjacency matrix indices (in COO format). See the documentation for single mode: https://graphneural.network/data-modes/#single-mode

Long fix: I will update the SingleLoader because it should return the edge features in the correct format, so what you're seeing is a plain bug. Thanks for spotting the issue, I will fix it asap. 

Cheers

**Reply by melheithem** [reply]
Hi,

I can see better now, I have made the short fix you suggest and it is working fine for me.

Thank you so much !
